drivers/sdhc/sdhc_renesas_ra.c

```diff
@@ -0,0 +1,718 @@
+{
+	int fsp_err = 0;
+
+	while (retries > 0) {
+		fsp_err = r_sdhi_command_send(&priv->sdmmc_ctrl, ra_cmd->opcode, ra_cmd->arg);
+		if (fsp_err != 0) {
+			retries--; /* error, retry */
+		} else {
+			break;
+		}
+	}
+	return err_fsp2zep(fsp_err);
+}
+
+/*
+ * Send CMD or CMD/DATA via SDHC
+ */
+static int sdhc_ra_request(const struct device *dev, struct sdhc_command *cmd,
+			   struct sdhc_data *data)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	int retries = (int)(cmd->retries + 1); /* first try plus retries */
+	fsp_err_t fsp_err = 0;
+	uint32_t timeout_cfg = 0;
+	int ret = 0;
+	sdmmc_priv_csd_reg_t p_csd_reg;
+
+	struct sdmmc_ra_command ra_cmd = {
+		.opcode = cmd->opcode,
+		.arg = cmd->arg,
+	};
+
+	if (data) {
+		ra_cmd.data = (uint8_t *)data->data;
+		ra_cmd.sector_start = data->block_addr;
+		ra_cmd.sector_count = data->blocks;
+		timeout_cfg = data->timeout_ms;
+	} else {
+		timeout_cfg = cmd->timeout_ms;
+	}
+	/*
+	 * Handle opcode with RA specifics
+	 */
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+	case SD_ALL_SEND_CID:
+	case SD_SEND_RELATIVE_ADDR:
+	case SD_SELECT_CARD:
+	case SD_SEND_IF_COND:
+	case SD_SET_BLOCK_SIZE:
+	case SD_ERASE_BLOCK_START:
+	case SD_ERASE_BLOCK_END:
+	case SD_ERASE_BLOCK_OPERATION:
+	case SD_APP_CMD:
+	case SD_SEND_STATUS:
+		/* Send command with argument */
+		ret = sdhc_ra_send_cmd(priv, &ra_cmd, retries);
+		if (ret < 0) {
+			goto end;
+		}
+		break;
+	case SD_SEND_CSD:
+		/* Read card specific data register */
+		ret = sdhc_ra_send_cmd(priv, &ra_cmd, retries);
+		if (ret < 0) {
+			goto end;
+		}
+		/* SDResponseR2 are bits from 8-127, first 8 MSBs are reserved */
+		p_csd_reg.reg.sdrsp10 = priv->sdmmc_ctrl.p_reg->SD_RSP10;
+		p_csd_reg.reg.sdrsp32 = priv->sdmmc_ctrl.p_reg->SD_RSP32;
+		p_csd_reg.reg.sdrsp54 = priv->sdmmc_ctrl.p_reg->SD_RSP54;
+		p_csd_reg.reg.sdrsp76 = priv->sdmmc_ctrl.p_reg->SD_RSP76;
+
+		/* Get the CSD version. */
+		uint32_t csd_version = p_csd_reg.csd_v1_b.csd_structure;
+		uint32_t mult;
+
+		if ((SDHI_PRV_CSD_VERSION_1_0 == csd_version) ||
+		    (SDMMC_CARD_TYPE_MMC == priv->sdmmc_ctrl.device.card_type)) {
+			mult = (1U << (p_csd_reg.csd_v1_b.c_size_mult + 2));
+			priv->sdmmc_ctrl.device.sector_count =
+				((p_csd_reg.csd_v1_b.c_size + 1U) * mult);
+
+			/* Scale the sector count by the actual block size. */
+			uint32_t read_sector_size = 1U << p_csd_reg.csd_v1_b.read_bl_len;
+
+			priv->sdmmc_ctrl.device.sector_count =
+				priv->sdmmc_ctrl.device.sector_count *
+				(read_sector_size / SDHI_MAX_BLOCK_SIZE);
+
+			if (SDMMC_CARD_TYPE_MMC == priv->sdmmc_ctrl.device.card_type) {
+				/*
+				 * If c_size is 0xFFF, then sector_count should be obtained from the
+				 * extended CSD. Set it to 0 to indicate it should come from the
+				 * extended CSD later.
+				 */
+				if (SDHI_PRV_SECTOR_COUNT_IN_EXT_CSD == p_csd_reg.csd_v1_b.c_size) {
+					priv->sdmmc_ctrl.device.sector_count = 0U;
+				}
+			}
+		}
+
+#if SDHI_CFG_SD_SUPPORT_ENABLE
```

Comment:
Just a question- where does this definition come from, the HAL? Could we use 
```
if (IS_ENABLED(SDHI_CFG_SD_SUPPORT_ENABLE))
```
here, or would that cause a compile error?

Reply:
Yes. I tried to use if (IS_ENABLED(SDHI_CFG_SD_SUPPORT_ENABLE)) but it causes a compile error.